plugin/compat/v1.go

```diff
@@ -0,0 +1,183 @@
+package compat
+
+import (
+	"net/url"
+
+	"github.com/gin-gonic/gin"
+	papiv1 "github.com/gotify/plugin-api"
+)
+
+// PluginV1 is an abstraction of a plugin written in the v1 plugin API. Exported for testing purposes only.
+type PluginV1 struct {
+	Info        papiv1.Info
+	Constructor func(ctx papiv1.UserContext) papiv1.Plugin
+}
+
+// APIVersion returns the API version
+func (c PluginV1) APIVersion() string {
+	return "v1"
+}
+
+// PluginInfo implements conpat/Plugin
+func (c PluginV1) PluginInfo() Info {
+	return Info{
+		Version:     c.Info.Version,
+		Author:      c.Info.Author,
+		Name:        c.Info.Name,
+		Website:     c.Info.Website,
+		Description: c.Info.Description,
+		License:     c.Info.License,
+		ModulePath:  c.Info.ModulePath,
+	}
+}
+
+// NewPluginInstance implements compat/Plugin
+func (c PluginV1) NewPluginInstance(ctx UserContext) PluginInstance {
+	instance := c.Constructor(papiv1.UserContext{
+		ID:    ctx.ID,
+		Name:  ctx.Name,
+		Admin: ctx.Admin,
+	})
+
+	compat := &PluginV1Instance{
+		instance: instance,
+	}
+
+	if displayer, ok := instance.(papiv1.Displayer); ok {
+		compat.displayer = displayer
+	}
+
+	if messenger, ok := instance.(papiv1.Messenger); ok {
+		compat.messenger = messenger
+	}
+
+	if configurer, ok := instance.(papiv1.Configurer); ok {
+		compat.configurer = configurer
+	}
+
+	if storager, ok := instance.(papiv1.Storager); ok {
+		compat.storager = storager
+	}
+
+	if webhooker, ok := instance.(papiv1.Webhooker); ok {
+		compat.webhooker = webhooker
+	}
+
+	return compat
+}
+
+// PluginV1Instance is an adapter for plugin using v1 API
+type PluginV1Instance struct {
+	instance   papiv1.Plugin
+	messenger  papiv1.Messenger
+	configurer papiv1.Configurer
+	storager   papiv1.Storager
+	webhooker  papiv1.Webhooker
+	displayer  papiv1.Displayer
+}
+
+// DefaultConfig see papiv1.Configurer
+func (c *PluginV1Instance) DefaultConfig() interface{} {
+	if c.configurer != nil {
+		return c.configurer.DefaultConfig()
+	}
+	return struct{}{}
+}
+
+// ValidateAndSetConfig see papiv1.Configurer
+func (c *PluginV1Instance) ValidateAndSetConfig(config interface{}) error {
+	if c.configurer != nil {
+		return c.configurer.ValidateAndSetConfig(config)
+	}
+	return nil
+}
+
+// GetDisplay see papiv1.Displayer
+func (c *PluginV1Instance) GetDisplay(location *url.URL) string {
+	if c.displayer != nil {
+		return c.displayer.GetDisplay(location)
+	}
+	return ""
+}
+
+// SetMessageHandler see papiv1.Messenger
+func (c *PluginV1Instance) SetMessageHandler(h MessageHandler) {
+	if c.messenger != nil {
+		c.messenger.SetMessageHandler(&PluginV1MessageHandler{WrapperHandler: h})
+	}
+}
+
+// RegisterWebhook see papiv1.Webhooker
+func (c *PluginV1Instance) RegisterWebhook(basePath string, mux *gin.RouterGroup) {
+	if c.webhooker != nil {
+		c.webhooker.RegisterWebhook(basePath, mux)
```

Comment:
I kinda don't like, that the webhook will still be registed even if the plugin is disabled.

Reply:
We could add a middleware to the plugin group which returns 400 bad request if the plugin is disabled.

Reply:
Then there would be too much burden on the database if the webhooks are called at a really high frequency. Maybe use the [callback](http://gorm.io/docs/write_plugins.html) function on the `GormDatabse` to provide notifications? (actually I was also wondering why we need to implement notifications for UserAdded and UserRemoved manually).

Since this PR is really getting uncontrollably bigger, lets take these unnessesery changes later if you choose to migrate to gorm callbacks.

Reply:
I don't think the check for enabled is a problem for the database, on every other request the client/apptoken will get queried for authentication.

Yeah lets not use callbacks for now.